migrator.py

```diff
@@ -3,39 +3,153 @@
+
+    # Build a contact_id to name(s) dictionary
+    global_names: GlobalNames = {}
+
+    if contacts_file is not None:
+        # contacts.xml seems more authoratative than *.ini files
+        # Use this whenever possible.
+        # Can be found at %LocalAppData%\Google\Picasa2\contacts\contacts.xml
+        # Also can be obtained from performing a backup operation from 
+        # within Picasa.  It will be in the backup location as 
+        # $Application Data\Google\Picasa2\contacts\backup.xml
+        logging.info(f'Reading contacts from {contacts_file}')
+        tree = ET.parse(contacts_file)
+        for contact in tree.getroot():
+            logging.info(f'{contact.attrib}')
+            global_names[contact.attrib['id']] = contact.attrib['name']
+
+    else:
+        # Alternate way using *.ini files
+        # Sometimes names will be missing or conflicting if found 
+        # different in multiple directories.  Conflicting names
+        # will be merged by this script by concatenating them with " | " 
+        # separators in sorted order.  Missing names will have an
+        # auto generated value of  
+        contact_id_2_nameset: Dict[str, set] = {}
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+        for input_dir, subdirs, files in os.walk(input_root_dir):
+            for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
+                learn_old_contact_ids(input_dir,ini_file,contact_id_2_nameset)
+
+        logging.info("contact_id_2_nameset")
+        logging.info("------------------------------------------------------------")
+        try:
+            for contact_id, names in contact_id_2_nameset.items():
+                person_name = ""
+                for i, item in enumerate(sorted(names)):
+                    person_name += item
+                    if i < (len(names)-1):
+                        # ambiguous contact_id will have a concatenated list of names
+                        person_name += "|"
+                logging.info(f"'{contact_id}': '{person_name}'")
+                # Create or add contacts
+                global_names[contact_id] = person_name
+        except Exception as e:
+            logging.warning(f"Exception: {e}")
+        logging.info("------------------------------------------------------------")
+
     contact_tags_per_dir: Dict[Path, ContactTags] = {}
-    for input_dir, subdirs, files in os.walk(input_root_dir):
-        dir = Path(input_dir)
-        if _PICASA_INI_FILE in files:
-           contact_tags_per_dir[dir] = migrate_directory(dir, files, db, contact_tags_per_dir, dry_run=dry_run)
-        elif any([_is_photo_file(file) for file in files]):
-            logging.warning('Found photos but no %s in %s' % (_PICASA_INI_FILE, dir))
 
+    for ini_file in (_PICASA_INI_FILE,_OLD_PICASA_INI_FILE):
```

Comment:
This outer loop looks odd.

Reply:
Maybe we need sth like `if file.exists(_PICASA_INI_FILE)`?

Reply:
I think I'll use the file.exists() check and remove the loop.  Seems like you get only .picasa.ini or Picasa.ini but not both -- at least in my collection.